src/osd/PG.cc

```diff
@@ -271,6 +331,7 @@ PG::PG(OSDService *o, OSDMapRef curmap,
 
 PG::~PG()
 {
+  pgstate_history.set_pg_in_destructor();
```

Comment:
After looking into this and discussions with Josh we decided that changing the destructor-related code is probably not worth the extra effort involved (it's tricky due to object lifetimes).